golang/pkg/rnr/tasks.go

```diff
@@ -28,10 +33,84 @@ func taskSchedState(pbt *pb.Task) TaskState {
 	return UNKNOWN
 }
 
+// TaskCallback is installed by higher-level construct, such as callback task or nested task.
+type TaskCallback func(context.Context, *Task)
+type StateUpdateCallback func(*pb.Task) *pb.Task
+
 // Task is a generic interface for pollable tasks
-type Task interface {
-	Poll()
-	Proto(updater func(*pb.Task)) *pb.Task
-	SetState(pb.TaskState)
-	GetChild(name string) Task
+type Task struct {
+	cb           TaskCallback
+	pb           *pb.Task
+	children     []*Task
+	has_children bool
+}
+
+func NewTask(name string, children bool, cb TaskCallback) *Task {
+	return &Task{
+		cb: cb,
+		pb: &pb.Task{
+			Name:  name,
+			State: pb.TaskState_PENDING,
+		},
+		children:     []*Task{},
+		has_children: children,
+	}
+}
+
+func (task *Task) Poll() {
+	task.cb(context.TODO(), task)
+}
+
+func (task *Task) Proto(updater StateUpdateCallback) *pb.Task {
+
+	oldState, ok := proto.Clone(task.pb).(*pb.Task)
+	if !ok {
+		log.Fatalf("Failed to clone proto")
+	}
+
+	if updater != nil {
+		task.pb = updater(oldState)
+	}
+
+	// Rebuild the children protobufs.
+	// This is terribly inefficient, but probably the easiest thing to do.
+	task.pb.Children = []*pb.Task{}
+	for _, c := range task.children {
+		p := c.Proto(nil)
+		task.pb.Children = append(task.pb.Children, p)
+	}
```

Comment:
If you care about efficiency, this should make things a bit more efficient, although not so sure with a small number of tasks:

```suggestion
	task.pb.Children = make([]*pb.Task, len(task.children))
	for i, c := range task.children {
		task.pb.Children[i] = c.Proto(nil)
	}
```

Basically you allocate an array of the right size just once instead of having the compiler doing bound checks and regrowing the slice on each `append`.

Reply:
Yup, definitely a good idea. :)